src/main/haskell/kore/src/Kore/Step/SmtLemma.hs

```diff
@@ -0,0 +1,182 @@
+    mapM_ declareSymbol (indexedModuleObjectSymbolSentences m)
+    mapM_ declareRule (indexedModuleAxioms m)
+  where
+    declareSort :: forall .
+        ( Given (MetadataTools Object StepperAttributes)
+        )
+        =>  ( StepperAttributes
+            , SentenceSort
+                Object
+                (KorePattern Domain.Builtin Variable (Unified (Valid (Unified Variable))))
+            )
+            -> SMT ()
+    declareSort (atts, _) =
+        case getSmtlib $ smtlib atts of
+            Just (SMT.List (SMT.Atom name : sortArgs)) -> do
+                _ <- SMT.declareSort name (length sortArgs)
+                pure ()
+            _ -> pure ()
+    declareSymbol :: forall .
+        ( Given (MetadataTools Object StepperAttributes)
+        )
+        =>  ( StepperAttributes
+            , SentenceSymbol
+                Object
+                (KorePattern Domain.Builtin Variable (Unified (Valid (Unified Variable))))
+            )
+            -> SMT (Maybe ())
+    declareSymbol (atts, symDeclaration) = runMaybeT $
+        case getSmtlib $ smtlib atts of
+            Just (SMT.List (SMT.Atom name : _)) -> do
+                inputSorts <- mapM translateSort (sentenceSymbolSorts symDeclaration)
+                resultSort <- translateSort (sentenceSymbolResultSort symDeclaration)
+                _ <- SMT.declareFun name inputSorts resultSort
+                pure ()
+            _ -> pure ()
+    declareRule :: forall sortParam .
+        ( Given (MetadataTools Object StepperAttributes)
+        )
+        =>  ( AxiomPatternAttributes
+            , SentenceAxiom
+                sortParam
+                (KorePattern Domain.Builtin Variable (Unified (Valid (Unified Variable))))
+            )
+            -> SMT (Maybe ())
+    declareRule (atts, axiomDeclaration) = runMaybeT $ do
+        guard (isSmtLemma $ smtLemma atts)
+        pat <- getRight
+            $ fromKorePattern Object
+            $ sentenceAxiomPattern axiomDeclaration
+        (lemma, vars) <-
+            runTranslator
+          $ translatePredicate translateUninterpreted
+          $ wrapPredicate
+          $ normalizeRule pat
+        SMT.assert (addQuantifiers vars lemma)
+
+    addQuantifiers vars lemma | null vars = lemma
+    addQuantifiers vars lemma = SMT.List
+        [ SMT.Atom "forall"
+        , SMT.List
+            [ SMT.List [sexpr, t] | (sexpr, t) <- Map.elems vars ]
+        , lemma
+        ]
+
+    translateSort
+        :: Given (MetadataTools Object StepperAttributes)
+        => Sort Object
+        -> MaybeT SMT SExpr
+    translateSort sort@(SortActualSort (SortActual _ children)) =
+        case getSmtlib $ smtlib $ sortAttributes given sort of
+            Just sExpr -> do
+                children' <- mapM translateSort children
+                pure $ applySExpr sExpr children'
+            Nothing -> mzero
+    translateSort _ = mzero
+
+getRight :: MonadPlus m => Either a b -> m b
```

Comment:
Why not `Alternative` instead of `MonadPlus` and `getRight = either (const empty) pure`?

Reply:
...because I forgot Alternative exists